streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java

```diff
@@ -265,23 +264,45 @@ private InMemoryKeyValueIterator(final Set<Bytes> keySet, final boolean forward)
 
         @Override
         public boolean hasNext() {
-            return iter.hasNext();
+            if (!open) {
+                throw new InvalidStateStoreException(String.format("Store %s has closed", name));
+            }
+            if (currentKey != null) {
+                if (map.containsKey(currentKey)) {
+                    return true;
+                } else {
+                    currentKey = null;
+                    return hasNext();
+                }
+            }
+            if (!iter.hasNext()) {
+                return false;
+            }
+            currentKey = iter.next();
+            return hasNext();
         }
 
         @Override
         public KeyValue<Bytes, byte[]> next() {
-            final Bytes key = iter.next();
-            return new KeyValue<>(key, map.get(key));
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            final KeyValue<Bytes, byte[]> ret = new KeyValue<>(currentKey, map.get(currentKey));
+            currentKey = null;
+            return ret;
         }
 
         @Override
         public void close() {
-            // do nothing
+            open = false;
```

Comment:
We should remove this as well when we remove the local `open` field in the iterator class so we don't accidentally close the store itself